### ISSUE TITLE
Should we support swizzles.

### ISSUE THREAD

**Issue by bit-hack** [8BitPimp/tack#3]
Should we support swizzles via some syntactic sugar?
Supporting all combinations I think ends up being 256 variants... :/
Can this be moved out to an enum and taken as a single template argument to shuffle?
The problem I see currently is that having four different lane indexes for shuffle can be a little confusing and hard to parse visually.
The same swizzle enum could be used later for int4 etc.

```
float4 a(.1f, .2f, .3f, .4f);
float4 b = a.xxyz();
```

or

```
float4 = my_vec.swizzle<e_xxyz>();
```

**Comment by bit-hack** [comment]
Certain swizzles can also be mapped directly to specific intrinsics, for instance:

```
// _mm_movelh()
float4 v = my_vec.swizzle<e_xyxy>();

// _mm_movehl()
float4 v = my_vec.swizzle<e_zwzw>();

// _mm_unpacklo_ps()
float4 v = my_vec.swizzle<e_xxyy>();

// _mm_unpackhi_ps()
float4 v = my_vec.swizzle<e_zzww>();
```

If we go down the route of abstracting these behind swizzling, how can we make sure that given a specific enum, the code will optimized down only to the correct intrinsic?

**Comment by bit-hack** [comment]
Would a switch case get optimized away cleanly?
Also, i'm making the possibly incorrect assumption that things like unpack and movelh would be faster...

```
template <swizzle_t SWIZZLE>
float4 swizzle() const {
  switch (SWIZZLE) {
  case (e_xxyy):
    return _mm_unpacklo_ps(m, m);
  default:
    return _mm_shuffle_ps(m, SWIZZLE);
  }
}
```

**Comment by kbenzie** [comment]
Yeah the shuffle as it is now isn't really safe because `uint8_t` isn't limited to using 2 bits to select from the 4 channel indexes.

Swizzling in C++ annoys me because I'm used to the GLSL syntax where the swizzle looks like a data member of a class, would be nice to have C# style attribute member functions but anyway back to reality.

A swizzle function taking an enum as the template parameter is really easy to ensure its optimized to the correct intrinsic as long as you fully specialize the function.

Using a switch statement is interesting, I have a gut feeling it might generate weird warnings but since its a template funciton it might actually optimize away nice a cleanly to the specific case since its known at compile time. Will have to test this one.

**Comment by bit-hack** [comment]
#7 should address this issue.

**Comment by kbenzie** [comment]
Closed by #7 .
